libs/framework/include/celix/dm/ServiceDependency_Impl.h

```diff
@@ -403,22 +531,64 @@ void ServiceDependency<T,I>::setupCallbacks() {
     int(*cadd)(void*, void *, const celix_properties_t*) {nullptr};
     int(*crem)(void*, void *, const celix_properties_t*) {nullptr};
 
-    if (setFp != nullptr) {
-        cset = [](void* handle, void *service, const celix_properties_t* props) -> int {
+    if (setFp || setFpUsingSharedPtr) {
+        cset = [](void* handle, void* rawSvc, const celix_properties_t* rawProps) -> int {
+            int rc = 0;
             auto dep = (ServiceDependency<T,I>*) handle;
-            return dep->invokeCallback(dep->setFp, props, service);
+            if (dep->setFp) {
+                rc = dep->invokeCallback(dep->setFp, rawProps, rawSvc);
+            }
+            if (dep->setFpUsingSharedPtr) {
+                auto svcId = dep->setService.second ? dep->setService.second->getAsLong(celix::SERVICE_ID, -1) : -1;
+                std::weak_ptr<I> currentSvc = dep->setService.first;
+                std::weak_ptr<const celix::Properties> currentProps = dep->setService.second;
+                auto svc = std::shared_ptr<I>{static_cast<I*>(rawSvc), [](I*){/*nop*/}};
+                auto props = rawProps ? celix::Properties::wrap(rawProps) : nullptr;
+                dep->setService = std::make_pair(std::move(svc), std::move(props));
+                dep->setFpUsingSharedPtr(dep->setService.first, dep->setService.second);
+                dep->waitForExpired(currentSvc, svcId, "service pointer");
```

Comment:
Am I correct in understanding that this call will hang until the service to which it is being added is completely stopped? I don't see another time when the weak_ptrs expire.

Reply:
The currentSvc is the service "currently" set and which just has been replaced (Maybe rename it so replacedSvc?).
So that waitForExpired will wait until the user has no instance of that service anymore. 

With raw pointer this check is not possible, but we assume that the user will not use the old service anymore after a new service is set. With shared_ptr this can be checked.

Reply:
Does the user stuff happen on another thread then?